beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1725,13 +1733,63 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         result
     }
 
+    fn prune_abandoned_forks(
+        &self,
+        old_finalized_block: &Hash256,
+        new_finalized_block: &Hash256,
+    ) -> Result<(), Error> {
+        let old_finalized_slot = self
+            .get_block(old_finalized_block)?
+            .ok_or_else(|| Error::MissingBeaconState(*old_finalized_block))?
+            .message
+            .slot;
+
+        let new_finalized_slot = self
+            .get_block(new_finalized_block)?
+            .ok_or_else(|| Error::MissingBeaconState(*new_finalized_block))?
+            .message
+            .slot;
+
+        // Collect hashes from new_finalized_block up to old_finalized_block
+        let newly_finalized_blocks: HashSet<Hash256> = HashSet::from_iter(
+            ParentRootBlockIterator::new(&*self.store, *new_finalized_block)
+                .take_while(|(block_hash, _)| block_hash != old_finalized_block)
+                .map(|(block_hash, _)| block_hash),
+        );
+
+        // Forget blocks with the slot number in the range [old_finalized_slot, new_finalized_slot]
+        // that are not in the newly_finalized_blocks.
+        for (head, _) in self.heads() {
+            if head == *new_finalized_block {
+                continue;
```

Comment:
This _should_ be impossible to reach but I'm not opposed to having it here :)

Reply:
Doesn't `self.head_tracker` include `canonical_head`?

Reply:
It does, but the finalized block necessarily has at least 2 epochs of chain following it, and one of those later blocks will be the head, _not_ the finalized block (a block can never be the head and be finalized).